File: stacks/laravel/app/Http/Controllers/Billing/DmController.php
<?php

namespace App\Http\Controllers\Billing;

use App\Http\Controllers\Controller;
use App\Models\Prevention;
use App\Support\FixedData;
use Illuminate\Contracts\View\View;
use Illuminate\Database\Eloquent\Collection;
use Illuminate\Http\Request;
use Symfony\Component\HttpFoundation\StreamedResponse;

/**
 * DM管理（画面16）。契約: spec/openapi.yaml `/dm` `/dm.csv`。
 * 画面とCSVは**同じ絞り込みロジック**（query()）を使い、件数が食い違わないようにする。
 */
class DmController extends Controller
{
    public function index(Request $request): View
    {
        $rows = $this->query($request);

        return view('billing.dm', [
            'rows' => $rows,
            'kinds' => FixedData::master('prevention_kinds'),
            'type' => $request->query('type'),
            'field' => $request->query('field', 'next_due_date'),
            'from' => $request->query('from'),
            'to' => $request->query('to'),
        ]);
    }

    public function csv(Request $request): StreamedResponse
    {
        $rows = $this->query($request);

        return response()->streamDownload(function () use ($rows) {
            $out = fopen('php://output', 'w');
            fputcsv($out, ['karte_no', 'owner_name_kanji', 'patient_name_kanji', 'kind', 'next_due_date', 'performed_date']);
            foreach ($rows as $r) {
                fputcsv($out, [
                    $r->patient->karte_no,
                    $r->patient->owner->name_kanji,
                    $r->patient->name_kanji,
                    $r->kind,
                    optional($r->next_due_date)->toDateString(),
                    optional($r->performed_date)->toDateString(),
                ]);
            }
            fclose($out);
        }, 'dm.csv', ['Content-Type' => 'text/csv']);
    }

    /**
     * 画面・CSV共通の絞り込み。次回予定日が入っていない記録は出さない。
     * deleted_at が入っている Patient／Owner に紐づく記録は出さない（spec/screens.md画面16）。
     */
    private function query(Request $request): Collection
    {
        $field = in_array($request->query('field'), ['next_due_date', 'performed_date'], true)
            ? $request->query('field') : 'next_due_date';

        $query = Prevention::with(['patient.owner'])
            ->whereNotNull('next_due_date')
            ->whereHas('patient', fn ($q) => $q->whereNull('deleted_at'))
            ->whereHas('patient.owner', fn ($q) => $q->whereNull('deleted_at'));

        if ($request->filled('type')) {
            $kinds = FixedData::master('prevention_kinds');
            $kindIndex = (int) $request->query('type');
            if (isset($kinds[$kindIndex])) {
                $query->where('kind', $kinds[$kindIndex]['code']);
            }
        }

        if ($request->filled('from')) {
            $query->whereDate($field, '>=', $request->query('from'));
        }
        if ($request->filled('to')) {
            $query->whereDate($field, '<=', $request->query('to'));
        }

        return $query->orderBy($field)->get();
    }
}

File: stacks/rails/app/controllers/dm_controller.rb
require "csv"

# DM管理。spec/screens.md #16。次回予定日をもとに、案内対象を検索・CSVへ書き出す。
# 画面（index）とCSV（csv）で同じ絞り込み・同じ並びを使う（openapi.yaml の要求）。
class DmController < ApplicationController
  def index
    @kinds = FixedData::Masters.prevention_kinds
    @field = params[:field] == "performed_date" ? "performed_date" : "next_due_date"
    @rows = query_rows.to_a
  end

  def csv
    rows = query_rows.to_a
    csv_body = CSV.generate do |csv|
      csv << %w[karte_no owner_name patient_name kind next_due_date performed_date]
      rows.each do |r|
        csv << [r.patient.karte_no, r.patient.owner.name_kanji, r.patient.name_kanji,
                 r.kind, r.next_due_date, r.performed_date]
      end
    end
    send_data csv_body, type: "text/csv; charset=utf-8", filename: "dm.csv", disposition: "inline"
  end

  private

  # 次回予定日が入っていない記録・削除済みの患者/飼主に紐づく記録は対象外
  # （spec/screens.md #16「満たすべきこと」）。
  def query_rows
    field = params[:field] == "performed_date" ? "performed_date" : "next_due_date"

    scope = Prevention.joins(patient: :owner)
                       .where(patients: { deleted_at: nil }, owners: { deleted_at: nil })
                       .where.not(next_due_date: nil)

    if params[:type].present?
      code = FixedData::Masters.kind_code_for(params[:type])
      scope = scope.where(kind: code) if code
    end

    from = parse_date(params[:from])
    to = parse_date(params[:to])
    scope = scope.where(field => from..) if from
    scope = scope.where(field => ..to) if to

    scope.includes(patient: :owner).order(field => :asc)
  end

  def parse_date(value)
    return nil if value.blank?

    Date.parse(value)
  rescue ArgumentError
    nil
  end
end
